File: Part 1/java_solution/src/com/geh/Exchange/Exchange.java
package com.geh.Exchange;

import java.util.Map;
import java.util.HashMap;

public class Exchange {

    public Trade[] execute(Order order)
    {
        var orderBook = orderBooks.get(order.getInstrument());

        if (orderBook == null) {
            orderBook = new OrderBook();
            orderBooks.put(order.getInstrument(), orderBook);
        }

        return orderBook.execute(order);
    }

    private Map<String, OrderBook> orderBooks = new HashMap<>();

}

File: part_2/csharp_solution/Exchange/Exchange.cs
using System;
using System.Collections.Generic;

namespace Exchange;

public class Exchange
{
    public Trade[] Execute(Order order)
    {
        if (!OrderBooks.TryGetValue(order.Instrument, out var orderBook)) {
            orderBook = new OrderBook();
            OrderBooks[order.Instrument] = orderBook;    
        }    
        return orderBook.Execute(order);
    }

    Dictionary<string, OrderBook> OrderBooks { get; } = new Dictionary<string, OrderBook>();
}
